### PR TITLE
Physical Memory Protection (32/64) Tests and Covergroups

### PR DESCRIPTION
**Pull request by MuhammadHammad001**
<FOR DOC UPDATES FILL ONLY DESCRIPTION AND RELATED ISSUES SECTION AND REMOVE THE OTHERS>

## Description

>This PR adds the arch tests for the PMP for 32 bit and 64 bit architecture and also the Covergroups have been added as well for the coverage analysis. This implements the Physical Memory Protection (PMP) section from the following [Test Plan](https://docs.google.com/spreadsheets/d/1p13gic7BD6aq7n_dHrqti4QmlGpxY7FkF17RVbG4DC0/edit#gid=0). The assembly arch tests have been developed by [Mr. Umer Shahid](https://github.com/UmerShahidengr) and the coverpoints are developed by [Muhammad Hammad Bashir](https://github.com/MuhammadHammad001).

### Related Issues

> The links to the PRs that need to be merged before this PR:
> - [PR #110 -- Support for Translator and Macros in RISCOF](https://github.com/riscv-software-src/riscof/pull/110)
> - [PR #80 -- Support for Privileged Architecture in RISC-V ISAC](https://github.com/riscv-software-src/riscv-isac/pull/80)
### Ratified/Unratified Extensions

- [x] Ratified
- [ ] Unratified

### List Extensions

> Privileged Architecture -- Physical Memory Protection

### Reference Model Used

- [x] SAIL
- [ ] Spike
- [ ] Other - < SPECIFY HERE >

### Mandatory Checklist:

  - [x] All tests are compliant with the test-format spec present in this repo ?
  - [x] Ran the new tests on RISCOF with SAIL/Spike as reference model successfully ?
  - [x] Ran the new tests on RISCOF in [coverage mode](https://riscof.readthedocs.io/en/stable/commands.html#coverage)
  - [x] Link to Google-Drive folder containing the new coverage reports available at: < [Coverage Reports PMP](https://drive.google.com/drive/folders/1jJ7yJ9oaWcGO-rH0jh_qVnutv4gM7bdU) >
  - [x] Link to PR in RISCV-ISAC from which the reports were generated : < [PR #80 -- Support for Privileged Architecture in RISC-V ISAC](https://github.com/riscv-software-src/riscv-isac/pull/80)>
  - [x] Changelog entry created with a minor patch

### Optional Checklist:

  - [ ] RISCV-V CTG PR link if tests were generated using it : < SPECIFY HERE >
  - [x] Were the tests hand-written
  - [ ] Have you run these on any hard DUT model ? Please specify name and provide link if possible in the description
  - [ ] If you have modified arch\_test.h Please provide a detailed description of the changes in the Description section above.

### CODE REVIEW

**Comment by MuhammadHammad001** [comment]
@allenjbaum @jjscheel

File: CHANGELOG.md
Comment (jamesbeyond) [inline CHANGELOG.md:1]: please remove the updates in the CHANGELOG,  This is not required anymore, Maintainers will sort that out all together when making the release

**Comment by jamesbeyond** [comment]
@MuhammadHammad001 ,  could have look at the CI checks, seems those added tests are failing

**Comment by MuhammadHammad001** [comment]
@jamesbeyond The reason the CI is failing on the tests is because as mentioned in the description, these tests and covergroups are written following the new privilege architecture features support in RISC-V ISAC and RISCOF. Once, those PRs are merged, these tests will also pass the CI.

**Review by allenjbaum — APPROVED** [review]
I'm a little concerned that the helper macros are hardcoded as to the PMP registers numbers, Is there any way those can be passed in as a parameter? that will help other tests n the future

File: coverage/rv32_pmp.cgf [hunk 1]
```diff
@@ -0,0 +1,836 @@
+# This coverpoint checks the coverage of Lock bit test.
+# req = The old value of the pmpcfgs and pmpaddrs should be equal to the new one since the Lock bit is set and the new write try will fail.
+pmp_cfg_locked_write_unrelated:
+  config:
+    - check ISA:=regex(.*32.*); check ISA:=regex(.*I.*Zicsr.*);  def rvtest_mtrap_routine=True;
+  mnemonics:
+    "{csrrs, csrrw}" : 0
+  csr_comb:
+    # (Lock bit set) and (req) and (old_value_pmpcfg != 0)
+    (pmpcfg{{0 ... 15} >> 2} >> {0, 8, 16, 24}{[$2%4]} & 0x80 == 0x80) and (((old("pmpcfg$2") ^ pmpcfg$2) >> $2 & 0xFF) == 0x00) and old("pmpcfg$2") != 0: 0
```
Comment (allenjbaum) [inline coverage/rv32_pmp.cgf:10]: This applies to a lot of these cover points.
1:  0x80 is a constant. It should be named (e.g. pmp_lck_msk)
2: $1 is 0..15. $2 is 0000,1111,2222,3333
so selecting the shift  amount should be 0123,10123,0123,0123. That is simply {($1%4)<<3{  or {($1&4)<<3} ({0, 8, 16, 24}{[$2%4]} isn't wrong, but it is much harder to understand.  
But ((old("pmpcfg$2") ^ pmpcfg$2) >> $2 is wrong, it should be 
       ((old("pmpcfg$2") ^ pmpcfg$2) >> $3
 You might want to explain why it is important that the old value of pmpcfg!=0

**Review by allenjbaum — CHANGES_REQUESTED** [review]
Overall the test structure is for each mode, different RWX settings with R,W, & X ops (R,W,RW,RX,RWX) or priority cases with same settings & a lower priority with RWX settings, or
a 3 level priority with the middle being no allowed accesses. 
What is missing are for the first 2 cases: a noaccess case,  for the second RWX cases (which is OK because its redundant).
In all case tests, the restricted case is higher priority than the all or none case. 
I think you also need need tests where a restricted case is lower priority than no_Access or all access cases (see also comments on the first 2 tests). 
I don't think there are tests  here where the access is not matched by any entry, or paritally matched by any entry (without being unaligned). I think you can only do that with NA4 &ld, sd for RV64, and FLD,FSD in RV32 (so F-ext must be supported in that case)

File: coverage/rv32_pmp.cgf (hunk 1 above)
Comment (UmerShahidengr) [reply coverage/rv32_pmp.cgf]: Resolved by the latest commit

**Review by allenjbaum — DISMISSED** [review]
Bug is fixed

**Review by allenjbaum — APPROVED** [review]
The RV64 bugs fixed

**Comment by UmerShahidengr** [comment]
@allenjbaum please approve this PR again, your previous approval was dismissed because of the new commits, now it is passing the CI, so it is good to go now.

**Comment by jordancarlin** [comment]
Looks like the riscof_work folder was added in the latest commit. This definitely shouldn’t be included.

**Review by allenjbaum — APPROVED** [review]
CI is fixed, this is ready to go

**Pull request merged**